### PR TITLE
Add SetupNugetSources script

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: https://github.com/dotnet/arcade/issues/4250

Script used to add credentials configuration to NuGet.config when interacting with AzDO private feeds.

Some examples of the script input->output are here:

- [AspNetCore](https://dnceng.visualstudio.com/internal/_build/results?buildId=381926&view=logs&j=9895f7d8-b062-540b-461c-07f8695bfcc6&t=f957242c-0dcd-5deb-9038-96c336875764&l=15)
- [Toolset](https://dnceng.visualstudio.com/internal/_build/results?buildId=382016&view=logs&j=09811346-8274-5b72-2d96-1dd38f87c84b&t=caa591c7-4826-5622-1bf9-229e606b0124&l=15)
- [Core CLI](https://dev.azure.com/dnceng/internal/_build/results?buildId=382011&view=logs&j=09811346-8274-5b72-2d96-1dd38f87c84b&t=caa591c7-4826-5622-1bf9-229e606b0124&l=17)
- [Tooling](https://dnceng.visualstudio.com/internal/_build/results?buildId=382024&view=logs&j=7c8326b9-0a5f-532a-e6de-db8515c72d9a&t=1763a2c5-35ae-591a-54df-48990610746a&l=17)

Basically the script iterate over the package sources and for each `darc-int` source it will add a credential node for that feed name. Two other credential entries are added for the `dotnet3-internal` and `dotnet3-internal-transport` feeds.

### CODE REVIEW

**Comment by JohnTortugo** [comment]
Removed the YAML template as asked during last Tuesday meeting.

File: eng/common/SetupNugetSources.ps1 [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# This file is a temporary workaround for internal builds to be able to restore from private AzDO feeds.
+# This file should be removed as part of this issue: https://github.com/dotnet/arcade/issues/4080
+#
+# What the script does is iterate over all package sources in the pointed NuGet.config and add a credential entry
+# under <packageSourceCredentials> for each Maestro's managed private feed. Two additional credential
```
Comment (riarenas) [inline eng/common/SetupNugetSources.ps1]:
```suggestion
# under <packageSourceCredentials> for each Maestro managed private feed. Two additional credential 
```

File: eng/common/SetupNugetSources.ps1 [hunk 2]
```diff
@@ -0,0 +1,107 @@
+# This file is a temporary workaround for internal builds to be able to restore from private AzDO feeds.
+# This file should be removed as part of this issue: https://github.com/dotnet/arcade/issues/4080
+#
+# What the script does is iterate over all package sources in the pointed NuGet.config and add a credential entry
+# under <packageSourceCredentials> for each Maestro's managed private feed. Two additional credential 
+# entries are also added for the two private static internal feeds: dotnet3-internal and dotnet3-internal-transport.
+#
+# This script needs to be called in every job that will restore packages and which the base repo has
```
Comment (riarenas) [inline eng/common/SetupNugetSources.ps1]: We should provide some guidance on the YAML changes that need to be made to call this in the appropriate place. Maybe a doc with some YAML snippets would be helpful. We only want this script to be used when needed, so we should add some instructions on how that is achieved.

File: eng/common/SetupNugetSources.ps1 [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+# private AzDO feeds in the NuGet.config.
+
+[CmdletBinding()]
+param (
+    [Parameter(Mandatory = $true)][string]$ConfigFile,
+    [Parameter(Mandatory = $true)][string]$Password
+)
+
+# Add source entry to PackageSources
+function AddPackageSource($sources, [string]$SourceName, [string]$SourceEndPoint, $creds, $Username, $Password) {
```
Comment (riarenas) [inline eng/common/SetupNugetSources.ps1]: Curious, why only add the type to `SourceName` and `SourceEndPoint` ?

File: eng/common/SetupNugetSources.ps1 [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	ForEach ($PackageSource in $maestroPrivateSources) {
+		Write-Host "`tInserting credential for Maestro's feed:" $PackageSource.Key
+		AddCredential -Creds $creds -Source $PackageSource.Key -Username $Username -Password $Password
+	}
+}
+
+# Load NuGet.config
+$doc = New-Object System.Xml.XmlDocument
+$filename = (Get-Item $ConfigFile).FullName
+$doc.Load($filename)
```
Comment (riarenas) [inline eng/common/SetupNugetSources.ps1]: What happens if an invalid path is passed? How does the error look?

File: eng/common/SetupNugetSources.sh [hunk 5]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
+
```
Comment (riarenas) [inline eng/common/SetupNugetSources.sh]: Should we add set -e so thatthis exits on first error? or are there any commands here where we actually expect a non-zero exit code as success?

File: eng/common/SetupNugetSources.sh [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+CredToken=$2
+NL='\n'
+TB='\t'
+
+if [[ `uname -s` == "Darwin" ]]; then
+	NL=$'\\\n'
+	TB=''
+fi
+
+echo $OSTYPE
```
Comment (riarenas) [inline eng/common/SetupNugetSources.sh]: If this is more than a debug message, maybe the message needs  a little bit more context.

File: eng/common/SetupNugetSources.sh [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+uname -s
+
+# Ensure there is a <packageSources>...</packageSources> section.
+grep -i "<packageSources>" $ConfigFile
```
Comment (riarenas) [inline eng/common/SetupNugetSources.sh]: What happens if the ConfigFile path is incorrect?

File: eng/common/SetupNugetSources.sh [hunk 8]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
```
Comment (riarenas) [inline eng/common/SetupNugetSources.sh]: We should add the same disclaimers we have in the ps1 file to this one.

File: eng/common/SetupNugetSources.sh [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+		PackageSourceCredentialsNodeFooter="</packageSourceCredentials>"
+		NewCredential="${TB}${TB}<$FeedName>${NL}<add key=\"Username\" value=\"dn-bot\" />${NL}<add key=\"ClearTextPassword\" value=\"$CredToken\" />${NL}</$FeedName>"
+
+		sed -i.bak "s|$PackageSourceCredentialsNodeFooter|$NewCredential${NL}$PackageSourceCredentialsNodeFooter|" NuGet.config
+	fi
+done
+
+echo 
+echo
+cat $ConfigFile
```
Comment (riarenas) [inline eng/common/SetupNugetSources.sh]: I don't think we need to print the final file at the end.

File: eng/common/SetupNugetSources.ps1 [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+# Insert credential nodes for Maestro's private feeds
+InsertMaestroPrivateFeedCredentials -Sources $sources -Creds $creds -Password $Password
+
+AddPackageSource -Sources $sources -SourceName "dotnet3-internal" -SourceEndPoint "https://pkgs.dev.azure.com/dnceng/_packaging/dotnet3-internal/nuget/v2" -Creds $creds -Username "dn-bot" -Password $Password
+AddPackageSource -Sources $sources -SourceName "dotnet3-internal-transport" -SourceEndPoint "https://pkgs.dev.azure.com/dnceng/_packaging/dotnet3-internal-transport/nuget/v2" -Creds $creds -Username "dn-bot" -Password $Password
+
+$doc.Save($filename)
+
+Write-Host
+Get-Content -Path $ConfigFile
```
Comment (riarenas) [inline eng/common/SetupNugetSources.ps1]: I don't think it's necessary to print the file with the password to the console

File: eng/common/SetupNugetSources.ps1 (hunk 4 above)
Comment (JohnTortugo) [reply eng/common/SetupNugetSources.ps1]: Adjusted. Going up in next push.

File: eng/common/SetupNugetSources.sh (hunk 7 above)
Comment (JohnTortugo) [reply eng/common/SetupNugetSources.sh]: Adjusted. Going up in next push.

**Comment by JohnTortugo** [comment]
Tested the changes here: https://dnceng.visualstudio.com/internal/_build/results?buildId=409980&view=results and all looked good.
